src/bigint.rs

```diff
@@ -1850,8 +1872,14 @@ impl<'a, 'b> Rem<&'b BigInt> for &'a BigInt {
 
     #[inline]
     fn rem(self, other: &BigInt) -> BigInt {
-        let (_, r) = self.div_rem(other);
-        r
+        if let Some(other) = other.to_u32() {
+            self % other
+        } else if let Some(other) = other.to_i32() {
```

Comment:
Since `other` is unsigned (because it is a `BigInt`), how could `other.to_u32()` return `None` but `other.to_i32()` return `Some(_)`?

Reply:
i.e. I think you can delete lines 1877 and 1878

Reply:
Are you thinking of `BigUint`? `BigInt` can be negative...

Reply:
Yeah, you're right. I even retyped `BigInt` and it didn't click for me that that type is signed.

Since we're discussing this function, I suspect many instantiated `BigInt`s (in the wild) will be small and further, many will be bigger than one word. My point is that the range `2.1 billion.. 4.3 billion` is probably pretty uncommon, so it might make sense to just drop `u32` branch entirely. Alternatively it might make sense to try to cast to bigdigit for the unsigned branch (is there an easy way to do that?)

Reply:
The conversion checks should be relatively trivial, compared to the actual division. `Rem<i32>` even forwards to `Rem<u32>` with added sign handling, which is why it seems to make sense to deal with `u32` directly first. I don't feel strongly, but it feels like a micro-optimization to worry about this much, where the real win is just in dividing by a primitive.

Maybe you'd like to focus on this for a followup performance PR?

Reply:
Yeah, it would be silly to make this type of change without measuring. I am interested in doing some performance work, but I'm not exactly sure how to do it. It seems like num-bigint has used criterion.rs at some point? but I couldn't find it in the git log. I think criterion would probably be very helpful for optimizing.